gemini/use-cases/code/code_scanning_and_vulnerability_detection_with_geminipro_1.5.ipynb

```diff
@@ -0,0 +1,999 @@
+    "\n",
+    "auth.authenticate_user()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "x4A6vhaKQqKT",
+   "metadata": {
+    "id": "x4A6vhaKQqKT"
+   },
+   "source": [
+    "### Define project information & initialize Vertex AI\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "-c8ZjhqIgA7K",
+   "metadata": {
+    "id": "-c8ZjhqIgA7K"
+   },
+   "outputs": [],
+   "source": [
+    "# initialize variables\n",
+    "PROJECT_ID = \"your-project-id\"\n",
+    "REGION = \"us-central1\"\n",
+    "BUCKET_NAME = \"your-bucket-name\"\n",
+    "PREFIX = \"folder-name/\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "TwEE78OKgTRB",
+   "metadata": {
+    "id": "TwEE78OKgTRB"
+   },
+   "outputs": [],
+   "source": [
+    "# import and initialize Vertex AI\n",
+    "import vertexai\n",
+    "\n",
+    "vertexai.init(project=PROJECT_ID, location=REGION)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "zrgeiLeZRB8m",
+   "metadata": {
+    "id": "zrgeiLeZRB8m"
+   },
+   "source": [
+    "### Read the .py files from a GCS bucket and combines them into a single string.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "lZNJhxiignA5",
+   "metadata": {
+    "collapsed": true,
+    "id": "lZNJhxiignA5"
+   },
+   "outputs": [],
+   "source": [
+    "## This block of code reads the file from GCS bucket, iterates through those and combine them into a single string.\n",
+    "## it also Demarcate individual file content using the actual filename as a separator\n",
+    "## for example purposes we have just taken 10 files, however, this same process can scan 100s of files as well. specifically taking advantages of the long context window\n",
+    "## the sample files have been taken from an openly accessible Git repo-\n",
+    "## for your purpose, you can download from any similar or different repos. https://github.com/s2e-lab/SecurityEval/tree/main\n",
+    "\n",
+    "from google.cloud import storage\n",
+    "\n",
+    "\n",
+    "def process_py_files(bucket_name, prefix):\n",
+    "    \"\"\"\n",
+    "    Reads .py files from a GCS bucket and combines them into a single string.\n",
+    "    Returns:\n",
+    "      A string containing the combined content of all .py files.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    storage_client = storage.Client()\n",
+    "    bucket = storage_client.get_bucket(BUCKET_NAME)\n",
+    "    blobs = bucket.list_blobs(prefix=PREFIX)\n",
+    "\n",
+    "    combined_text = \"\"\n",
+    "    for blob in blobs:\n",
+    "        if blob.name.endswith(\".py\"):\n",
+    "            file_content = blob.download_as_string().decode(\"utf-8\")\n",
+    "            combined_text += f\"### File: {blob.name} ###\\n{file_content}\\n\"\n",
+    "\n",
+    "    return combined_text\n",
+    "\n",
+    "\n",
+    "combined_string = process_py_files(BUCKET_NAME, PREFIX)\n",
```

Comment:
Optional: You can actually load files direct from GCS into Gemini using the Vertex AI SDK.

https://github.com/googleapis/python-aiplatform/blob/88c6a6a4f11285d429c3777f59101e53e4672185/vertexai/generative_models/_generative_models.py#L1893C9-L1893C17

This _should_ work, I'm not sure if it can read the .py files as plain text.
```
Part.from_uri("file.py", mime_type="text/plain")
```